man/man1/send.1
.TH SEND 1 "4 January 1986"
.SH NAME
send \- send mail
.SH SYNOPSIS
.B send
[address\.\.\.] [\-a\ field:body] [\-b\ address\.\.\.]
[\-c\ address\.\.\.] [\-h\ host] [\-n] [\-s\ subject] [\-t\ address\.\.\.]
.SH DESCRIPTION
.PP
.I Send
sends messages to a user-specified list of recipients.
When invoked without arguments, it prompts for
the following information:
.sp 1
.RS
.nf
To:
cc:
Subject:
.fi
.sp 1
.RE
The expected user responses are:
.RS
.IP "1)" 4
To the "To:" and "cc:" prompts,
a list of valid user names (see below).
.IP "2)" 4
To the "Subject:" prompt,
a brief description of the subject of the message.
.RE
.PP
Copies of the message are delivered to recipients on
the "To:" and "cc:" lists.
A response to the "To:" prompt is required, unless blind carbon
copies (bcc) are specified.
A field may be left empty by typing a carriage return <CR>.
.PP
If the user's input for any of the above fields will require
more than one line, then when typing the input the user
should precede carriage
returns with a backslash (\\).  In any case, the user's input for any
one of these fields must not exceed 512 characters.
.sp
.SS Arguments
.PP
.I "Send"
may be invoked with arguments.
If they are specified, then the
contents of any resulting fields are displayed, before the
user is prompted for additional input.
Unless switches (a character preceded by a dash) are used, the
arguments are taken to be addresses for the "To:" field.
.RS
.IP "\-a" 4
causes the next argument to be added to the header of the 
message.  For example, ``\-a Reply-To:dcrocker''.  Enclose the
argument in quotes if it contains spaces.
.IP "\-b" 4
causes following arguments to be added to the "BCC:" field.
.IP "\-h" 4
causes the next argument to be used as the default host reference,
for addresses not explicitly containing a hostname.  When not
specified, the default is the local host.
.IP "\-c" 4
causes following arguments to be added to the "cc:" field.
.IP "\-s" 4
causes the next argument to be used as
the "Subject:" field's contents.
.IP "\-t" 4
causes following arguments to be added to the "To:" field.
.IP "\-d" 4
causes the state of the directedit
option to be changed (see "directedit" below).
Alone, "\-d" disables the directedit option.  "\-de" causes the directedit
option to use the \fIline editor\fR, and "\-dv" causes the directedit option
to use the \fIvisual (screen) editor\fR.
.IP "\-n" 4
prevents send from prompting for message text either directly or via
direct edit.  The user is immediately given the command prompt after
entering the header data.  Prompting for header data can be prevented
by giving null header arguments to send without contents.  For example:
.sp
.ti +.5i
\fIsend \-t "" \-c "" \-s "" \-n\fR
.RE
.sp
.SS "Valid User Addresses"
.PP
A valid address name takes one of the following forms:
.RS
.IP "1)" 4
A local Unix user's login name, such as "crocker."
.IP "2)" 4
A name which has been designated as a local alias, such as "support."
This usually will not look any different from a login name, except
that it may be longer than the standard 8-character limit.
.IP "3)" 4
The two-part "mailbox @ domain" address for a user on another machine.
.RE
.PP
Validation of names is done at posting time, and is performed
by the message delivery system, rather than by
.I "send."
All addresses
are made to have a domain reference.  If the user does not specify
a domain for an address, then the default host is assumed.
.PP
When the "To:", "cc:" and "Subject:" fields have been completed,
.I "send"
will prompt for the body of the message, with:
.sp 1
.ne 3
.RS
Type message; end with CTRL-D...
.RE
.PP
The user now enters the text of the message.
When the message is completed, or when the user wishes to insert
a file or make changes, s/he types a control-D at the beginning of a new line.
.I Send
then prompts the user for its next action.
.sp
.SS "User Options and Environmental Tailoring"
.PP
The user may define a file in his home directory called ".sendrc".  If such a
file exists,
.I "send"
will open the file and set options as specified,
otherwise the defaults will be used. Below are listed the options available
for user modification along with their default values.
.sp 2
.IP "aliases"
.RS
Allows the user to declare his own set of aliases for use with
.I "send"
and 
.I "resend."
Unless this file name begins with a '/', 
.I send
will look for this file in the user's home directory.
The format for this file is:
.RS
aliasname  user1@domain,user2@domain,user3@domain...
.RE
Note that each entry must be on a single line but each line may be
up to 511 characters long.
.RE
.IP "checker"
.RS
This will be the user's spelling checker while using \fIsend\fR.  The default
for this option is \fIspell\fR.
.RE
.IP "copyfile"
.RS
This is the file in which copies of messages will be stored if
requested by the user. Unless the first character of the supplied
filename is '/', the file will be created in the user's home
directory. The default for the copyfile is a file
called ".sent" in the users home directory.
See also 
"file, nofile, fileonquery" below.
.RE
.IP "directedit"
.RS
If this option is specified, the
.I send
program will immediately enter the editor after the "To:", "cc:" and "Subject:"
fields have been supplied.  The directedit option expects an argument
which specifies which editor to use.  "directedit v" causes the
visual editor to be used (see "veditor" below).  "directedit e" causes the
line editor to be used (see "editor" below).
.RE
.IP "draftdir"
.RS
If supplied, draft files will be created in this directory.
Otherwise, they will be created in the user's home directory.  
Draftdir should be a complete path name.
.RE
.IP "editor"
.RS
This will be the user's line editor while using 
.I "send."
The default for this option is system dependent, but normally set to
.I "ed."
If the environment variable ``EDITOR'' is set, it will be
taken as the default instead of the system default.
.RE 
.IP "file, nofile, fileonquery"
.RS
The presence of one of these options allows the users to control
the keeping of file copies.  ``File''
will result in a copy being kept of every
message sent by the user.  ``Nofile''
will turn off the file copy
option and no copies will ever be kept.  ``Fileonquery''
will cause
.I "send"
to query the user as to whether a copy should be kept or not.
.RE
.IP "header"
.RS
This option allows user-defined header lines to be added to the message being
sent.  The first argument is the field name (e.g. Organization).  The second
argument is the contents of the field (e.g. West Podunk College).  A separate
header line will be added for each "header" command in the .sendrc file.
.RE
.IP "paging"
.RS
Having this in your .sendrc will cause your message to be displayed a screenful
at a time when using the "review message" command.
.RE
.IP "signature, nosignature"
.RS
.I "Send"
will automatically fill in the ``From:'' field with your address.  The
form of the contents is either:
.RS
login@host  or  Full Name <login@host>
.RE
where ``login'' is your Unix login name, and ``host''
is the name of the Unix machine you are using.  The second form is the
default on
some systems.  On these systems, ``Full Name'' is derived from the system
password file.  The ``nosignature''
command forces \fIsend\fR to use the form:
.RS
login@host
.RE
.PP
You may personalize your messages by including a signature.  If
your 
.I ".sendrc"
file contains a signature line, its contents
will be used in the ``From:'' field, ahead of the address information.
The signature cannot contain any of the following characters
unless it is quoted with double quotes: `( ) < > @ , ; : \ \ " . [ ]' 
and it cannot be blank.
The program will verify the signature line for correctness at startup
and will complain if your signature is illegal.
The ``From:'' field then will appear as follows:
.RS
signature <login@host>
.RE
.RE
.IP "subargs"
.RS
Allows the user to specify additional flags for the \fIsubmit\fR program
('v' and 'm' are always passed). Probably not for general use.
.RE
.IP "veditor"
.RS
This is the \fIvisual editor\fR (screen editor) for use with
.I "send."
The default for
this editor is site dependent, but is usually some
version of vi or emacs.
If the environment variable ``VISUAL'' is set, it will be
taken as the default instead of the system default.
.RE
.sp
.PP
The format for a typical
.I ".sendrc"
profile is shown below:
.RS
.nf

copyfile sent_mail
draftdir .drafts
signature Joe Foobar
subargs w
veditor emacs
fileonquery
aliases joes_aliases
header Organization "West Podunk College"
.fi
.RE
.SS "Commands"
.PP
.I "Send"
commands must be followed by a carriage return.
Only enough of a command to make it unique need be typed.
The following are valid commands:
.IP "?"
.br
displays a list of
.I "send's"
commands.
.IP "bcc"
.br
prompts for addresses for the ``BCC:'' Blind carbon copies field.
If any addresses are specified for Blind carbon copies, 
they receive a slightly different message than those listed
in ``To:'' or ``cc:''.  Their copy has a header component
named ``BCC:''; its
contents are the string ``(Private)''.
If the message had no ``To:'' addresses,
then the ``To:'' field will contain the string ``list:'' and there will be no
BCC field.  In addition, the ``To:'' and ``cc:'' fieldnames
are modified to prevent
recipients from
automatically replying to
anyone in the ``To:'' or ``cc:'' fields.
.sp
.RS
NOTE: If a message has normal and BCC recipients and there is an error
with a BCC address, the normal addresses will have been processed,
and the copy of their message sent, before the BCC address error
is detected.
.RE
.IP "bye"
.br
exits the
.I "send"
program.
If the user has not yet sent the message, the user is asked
to confirm the action.
A ``yes'' (or 'y') is required, followed by a <CR>, or the command will
be aborted and the user returned to the ``Command or ?:'' prompt.
.IP "check spelling"
.br
will invoke the spelling checker with the draft filename as the argument
to the checker program.  (See ``checker'' user option above.)
.IP "delete body"
.br
permits the user to abort
.I "send"
without saving the draft body.
If \fIsend\fR is aborted without giving the ``delete body'' command,
a copy of the draft will be kept as drft.nnnnn in the draft directory
specified by the ``draftdir'' option above.
This command can also be given at any time to erase the current draft
before inputing more text.  It requires confirmation for obvious reasons.
.IP "edit"
.br
will invoke the user's line editor (usually the Unix editor \fIed\fR(1))
and read in the user's message body for editing.
(See ``editor'' user option above.)
.IP "file"
.br
indicates the user wishes to append a file to the end of the message body.
.I "Send"
will prompt for file name.
.IP "header edit"
.br
allows the user to change the contents of the header fields
(``To:'', ``cc:'', and ``Subject:'').  The ``BCC:'' field is included, if not
empty.
For each header field,
.I "send"
shows the current field contents and then repeats the prompt.
The user may type in new input which will then become the contents of
that header field.
To leave the contents of the field unchanged, the user types
only a carriage return.
To delete a specific entry, the user types a minus sign followed
by the address to be deleted (e.g. ``\-dcrocker@udel'').
To simply delete the first entry of a list, the
user can type a minus sign (\-) followed by a carriage return.
To delete the entire line, the user types a pound sign (#) followed by
a carriage return.
To add to the field, the user types a plus sign (+), followed
by the text to be added, followed by a carriage return. For example, to
add a name to a field, one would type the command ``+user@domain''.
More than one address can be added at a time by listing several addresses
separated by commas (e.g. +u1,u2).  To begin a line with one of these special 
editting characters (i.e. \-, +, or #), the user can precede it with a space.
.IP "input more body"
.br
this command allows the user to type more text onto the end of the
message body, ending with a control-d on a new line (as before).
.IP "post"
.br
same as ``send''.
.IP "quit"
.br
same as ``bye''.
.IP "review message"
.br
retypes message in its current form.
.IP "send"
.br
posts the message to the message delivery system.
The message is delivered immediately to local users and queued for later
delivery to non-local users.
The sender is informed of the posting disposition of the message.
Each address
is listed and followed by its status, indicating that the
address is ok, or that there was a problem.
.RS
.PP
If all addresses are valid, the sender is then informed that the message
has successfully been posted for delivery.
If any addresses are found to be invalid, the remaining addresses are
still checked, but submission is aborted.  The user
is returned to command input,
so that he may change the addresses and make them legal.
.RE
.IP "set"
.br
allows the user to check the current values of option variables. If followed
by arguments, allows the user to change the current setting of variables.
The changed value lasts only for the current invocation of 
.I "send."
Option values are those listed above under ``User Options and Environmental
Tailoring''.  For example, the command ``set editor ex'' would
change the ``editor'' option to ``ex'' for the duration of the current
.I send
session.
.IP "vedit"
.br
will invoke the screen editor as specified by the ``veditor''
option and read in the user's message body, for editing.
(See ``veditor'' user option above.)
.IP "program run"
.br
allows the user to give a shell command.
.I "Send"
prompts for program.
The user's response is passed to the shell.
Control is returned to
.I "send"
when the command has been executed.
.IP "CTRL-D"
.br
same as ``bye''
.SH FILES
.IP "<login directory>/.sendrc" 38
user options
.IP "<login directory>/aliases" 38
user-supplied alias file
.IP "<login directory>/.sent" 38
sent messages
.IP "<login directory>/drft.XXXXXX" 38
body of current message
.SH "SEE ALSO"
v6mail(1), msg(1), resend(1), snd(1)
.SH DIAGNOSTICS
All diagnostic messages are intended to be self explanatory.


Changes
Changes beta 3
	* Small bug in ap_lex.c hip@a.cs.okstate.edu
	* jromine@ics.uci.edu some changes for Solaris 2.5, Sun Pro
	  4.2 Compiler
	* linux 2.2.7, glibc2.1, egcs 2.91 (redhat 6.0)
	* sco (first attempt)
	* jromine@ics.uci.edu, boundary fix smtpsrvr
	* Small bug in submit/lnk_getaddr.c krueger@mathematik.uni-kl.de

Wed Dec 31 13:14:34 MET 1997	Kai Krueger <krueger@mathematik.uni-kl.de>

	* create file h/config.h.in and moved every define to it.

	* DOASSIGN renamed to HAVE_ASSIGN. Autodetection by configure

	* SECURETTY renamed to HAVE_SECURETTY, Autodetection by configure

	* NODIAL renamed to HAVE_DIAL. default is 'no'
	  may be enanble by configure --enable-dial
	  if compiling phone/smphone channel it wil be turned on by configure.

	* NAMESERVER renamed to HAVE_NAMESERVER. Autodetection by configure.
	  May be overwriten by --disable-nameserver
          - tb_ns.4.2.c and tb_ns.fake.c merged into one file tb_ns.c and
	  seperated by #ifdef HAVE_NAMESERVER...
	  - removed TB_NS in Makefile.com	  

	* HAVE_TCP_WRAPPER renamed to HAVE_LIBWRAP. Autodetection by configure.
	
	* <time.h> included in config.h. It's location is determined by
	  configure

	* determination of <dir.h>, <dirent.h>, a.s.o is done by configure
	  definition an include has moved to config.h

	* determination of index/rindex done by configure and moved to config.h

	* definition of sigtype renamed to RETSIGTYPE. Detection done by
	  configure.

	* return-type of sprintf determined by configure. Define BSD_SPRINTF
	  if char * is returned.

	* uip/Makefile, src/Makefile, lib/Makefile Makefile.com and Makefile
	  are generated by configure.

	* Value of CH_TB in Makefile.com set by configure. If you want
	  sequential tables run configure with the argument --disable-dbm

	* Added define HAVE_SGTTY_H for the existens of <sgtty.h>

	* Added check for <varargs.h> as define HAVE_VARARGS_H.
	  Added check for variable number of arguments if <varargs.h> don't
	  exists. If Compiler don't supports variable number of arguments
	  NO_VARARGS is defined. (NO_VARARGS should be renamed)
	  configure ist checking for both of it. But if <varagrs.h> isn't
	  found lib/util/multcat.c and lib/util/multcpy.c could't be compiled.
	  
	* test configure on machine that dosn't support multiple arguments.
	  (check of NO_VARARGS)

	* test for dup2() and fcntl(fd1, F_DUPFD, fd2) done.
	  NODUP2 replaced by HAVE_DUP2 and
	  NOFCTNL replaced by HAVE_FCNTL_F_DUPFD


uip/ucbmail/send.c
/*
 *  S E N D . C 
 *
 *  EE/CIS Computer Lab
 *  Department of Computer and Information Sciences
 *  Department of Electrical Engineering
 *  University of Delaware
 *
 *  REVISION HISTORY:
 *
 *  $Revision: 1.4 $
 *
 *  $Log: send.c,v $
 *  Revision 1.4  1986/01/14 16:29:24  galvin
 *  Initial revision
 *
 * Revision 1.4  86/01/14  16:29:24  galvin
 * Change send() to check for From line instead of believing it is there.
 * 
 * Change sendmail() to extract()/detract() the To field.
 * 
 * mail1() needed some tuning and cleaning.  Basically change it to
 * interface to the MMDF ml_* routines.  One other major change was to
 * take out the call to verify() and depend on MMDF to do it.  We have
 * to do this anyway since we are allowing address forms that choke Mail.
 * 
 * Change savemail() to include MMDF delimiters.
 * 
 * Revision 1.3  86/01/02  14:29:57  galvin
 * Add another parameter to send to indicate whether or not this message
 * should be delimited by MMDF message delimiters.
 * 
 * Revision 1.2  85/12/18  03:00:41  galvin
 * Added comment header for revision history.
 * 
 *
 */

/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
static char *sccsid = "@(#)send.c	5.2 (Berkeley) 6/21/85";
#endif not lint

#include "./rcv.h"
#include <ctype.h>
#include <sys/stat.h>

#include "./mmdf.h"

/*
 * Mail -- a mail program
 *
 * Mail to others.
 */

/*
 * Send message described by the passed pointer to the
 * passed output buffer.  Return -1 on error, but normally
 * the number of lines written.  Adjust the status: field
 * if need be.  If doign is set, suppress ignored header fields.
 * If delim is set, output MMDF delimiters around message.
 */
send(mailp, obuf, doign, delim)
	struct message *mailp;
	FILE *obuf;
	int delim;
{
	register struct message *mp;
	long c;
	FILE *ibuf;
	char line[LINESIZE], field[BUFSIZ];
	int lc, ishead, infld, fline, dostat;
	char *cp, *cp2;

	mp = mailp;
	ibuf = setinput(mp);
	c = mp->m_size;
	ishead = 1;
	dostat = 1;
	infld = 0;
	fline = 1;
	lc = 0;
	if (delim)
		fputs(delim1, obuf);
	while (c > 0L) {
		fgets(line, LINESIZE, ibuf);
		c -= (long) strlen(line);
		lc++;
		if (ishead) {
			/* 
			 * If first line is the "From line", then just print.
			 * Since we are dealing with MMDF, it probably isn't.
			 */
			if (fline) {
				fline = 0;
				if (!strncmp(line, "From ", 5))
				goto writeit;
			}
			/*
			 * If line is blank, we've reached end of
			 * headers, so force out status: field
			 * and note that we are no longer in header
			 * fields
			 */
			if (line[0] == '\n') {
				if (dostat) {
					statusput(mailp, obuf, doign);
					dostat = 0;
				}
				ishead = 0;
				goto writeit;
			}
			/*
			 * If this line is a continuation (via space or tab)
			 * of a previous header field, just echo it
			 * (unless the field should be ignored).
			 */
			if (infld && (isspace(line[0]) || line[0] == '\t')) {
				if (doign && isign(field)) continue;
				goto writeit;
			}
			infld = 0;
			/*
			 * If we are no longer looking at real
			 * header lines, force out status:
			 * This happens in uucp style mail where
			 * there are no headers at all.
			 */
			if (!headerp(line)) {
				if (dostat) {
					statusput(mailp, obuf, doign);
					dostat = 0;
				}
				putc('\n', obuf);
				ishead = 0;
				goto writeit;
			}
			infld++;
			/*
			 * Pick up the header field.
			 * If it is an ignored field and
			 * we care about such things, skip it.
			 */
			cp = line;
			cp2 = field;
			while (*cp && *cp != ':' && !isspace(*cp))
				*cp2++ = *cp++;
			*cp2 = 0;
			if (doign && isign(field))
				continue;
			/*
			 * If the field is "status," go compute and print the
			 * real Status: field
			 */
			if (icequal(field, "status")) {
				if (dostat) {
					statusput(mailp, obuf, doign);
					dostat = 0;
				}
				continue;
			}
		}
writeit:
		fputs(line, obuf);
		if (ferror(obuf))
			return(-1);
	}
	if (delim)
		fputs(delim2, obuf);
	if (ferror(obuf))
		return(-1);
	if (ishead && (mailp->m_flag & MSTATUS))
		printf("failed to fix up status field\n");
	return(lc);
}

/*
 * Test if the passed line is a header line, RFC 733 style.
 */
headerp(line)
	register char *line;
{
	register char *cp = line;

	while (*cp && !isspace(*cp) && *cp != ':')
		cp++;
	while (*cp && isspace(*cp))
		cp++;
	return(*cp == ':');
}

/*
 * Output a reasonable looking status field.
 * But if "status" is ignored and doign, forget it.
 */
statusput(mp, obuf, doign)
	register struct message *mp;
	register FILE *obuf;
{
	char statout[3];

	if (doign && isign("status"))
		return;
	if ((mp->m_flag & (MNEW|MREAD)) == MNEW)
		return;
	if (mp->m_flag & MREAD)
		strcpy(statout, "R");
	else
		strcpy(statout, "");
	if ((mp->m_flag & MNEW) == 0)
		strcat(statout, "O");
	fprintf(obuf, "Status: %s\n", statout);
}


/*
 * Interface between the argument list and the mail1 routine
 * which does all the dirty work.
 */

mail(people)
	char **people;
{
	register char *cp2;
	register int s;
	char *buf, **ap;
	struct header head;

	for (s = 0, ap = people; *ap != (char *) -1; ap++)
		s += strlen(*ap) + 1;
	buf = salloc(s+1);
	cp2 = buf;
	for (ap = people; *ap != (char *) -1; ap++) {
		cp2 = copy(*ap, cp2);
		*cp2++ = ' ';
	}
	if (cp2 != buf)
		cp2--;
	*cp2 = '\0';
	head.h_to = buf;
	head.h_subject = NOSTR;
	head.h_cc = NOSTR;
	head.h_bcc = NOSTR;
	head.h_seq = 0;
	mail1(&head);
}


/*
 * Send mail to a bunch of user names.  The interface is through
 * the mail routine below.
 */

sendmail(str)
	char *str;
{
	struct header head;

	if (blankline(str))
		head.h_to = NOSTR;
	else
		head.h_to = detract(elide(extract(str, GTO)), GTO);
	head.h_subject = NOSTR;
	head.h_cc = NOSTR;
	head.h_bcc = NOSTR;
	head.h_seq = 0;
	mail1(&head);
	return(0);
}

/*
 * Mail a message on standard input to the people indicated
 * in the passed header.  (Internal interface).
 */

mail1(hp)
	struct header *hp;
{
	register char *cp;
	int gotcha;
	struct name *to, *np;
	struct stat sbuf;
	FILE *mtf, *postage;
	int remote = rflag != NOSTR || rmail;

	/*
	 * Collect user's mail from standard input.
	 * Get the result as mtf.
	 */

	if ((mtf = collect(hp)) == NULL)
		return;
	hp->h_seq = 1;
	if (hp->h_subject == NOSTR)
		hp->h_subject = sflag;
	if (intty && value("askcc") != NOSTR)
		grabh(hp, GCC);
	else if (intty) {
		printf("EOT\n");
		fflush(stdout);
	}

	/*
	 * Now, get the user names from the combined to and cc lists
	 */

	senderr = 0;
	to = usermap(cat(extract(hp->h_bcc, GBCC),
	    cat(extract(hp->h_to, GTO), extract(hp->h_cc, GCC))));
	if (to == NIL) {
		printf("No recipients specified\n");
		goto topdog;
	}

	/*
	 * Look through the recipient list for names with /'s
	 * in them which we write to as files directly.
	 */

	to = outof(to, mtf, hp);
	if (senderr && !remote) {
topdog:
		rewind(mtf);
		if ((mtf = infix(hp, mtf)) == NULL) {
			fprintf(stderr, ". . . message lost, sorry.\n");
			return;
		}
		if (fsize(mtf) != 0) {
			remove(deadlettername);
			exwrite(deadlettername, mtf, 1);
		}
		goto out;
	}
	for (gotcha = 0, np = to; np != NIL; np = np->n_flink)
		if ((np->n_type & GDEL) == 0) {
			gotcha++;
			break;
		}
	if (!gotcha)
		goto out;
	to = elide(to);
	if (count(to) > 1)
		hp->h_seq++;
	hp->h_to = detract(to, GTO|GCOMMA);
	hp->h_cc = detract(to, GCC|GCOMMA);
	hp->h_bcc = detract(to, GBCC|GCOMMA);
	if (hp->h_seq > 0 && !remote) {
		if (fsize(mtf) == 0)
		    if (hp->h_subject == NOSTR)
			printf("No message, no subject; hope that's ok\n");
		    else
			printf("No message; hope that's ok\n");
		}
	if (debug) {
		printf("Recipients of message:\n");
		if (hp->h_to != NOSTR)
			printf(" \"%s\"", hp->h_to);
		if (hp->h_cc != NOSTR)
			printf(" \"%s\"", hp->h_cc);
		if (hp->h_bcc != NOSTR)
			printf(" \"%s\"", hp->h_bcc);
		printf("\n");
		fflush(stdout);
		return;
	}
	rewind(mtf);
	if (ml_init(YES, NO, NOSTR, hp->h_subject) != OK)
		goto topdog;
	if (!isnull(hp->h_to) && ml_adr(hp->h_to) != OK)
		goto topdog;
	if (!isnull(hp->h_cc) && ml_cc() != OK || ml_adr(hp->h_cc) != OK)
		goto topdog;
	if (!isnull(hp->h_bcc) && ml_adr(hp->h_bcc) != OK)
		goto topdog;
	if (ml_aend() != OK)
		goto topdog;
	if (ml_tinit() != OK)
		goto topdog;
	if (ml_file(mtf) != OK)
		goto topdog;
	if (ml_end(OK) != OK)
		goto topdog;
	rewind(mtf);
	if ((mtf = infix(hp, mtf)) == NULL) {
		fprintf(stderr, ". . . message lost, sorry.\n");
		return;
	}

	rewind(mtf);
	if ((cp = value("record")) != NOSTR)
		savemail(expand(cp), mtf);

		if (!stat(POSTAGE, &sbuf))
			if ((postage = fopen(POSTAGE, "a")) != NULL) {
				fprintf(postage, "%s %d %d\n", myname,
				    count(to), fsize(mtf));
				fclose(postage);
			}
out:
	fclose(mtf);
	return;
}

#ifdef NOT_USED
/*
 * Fix the header by glopping all of the expanded names from
 * the distribution list into the appropriate fields.
 * If there are any ARPA net recipients in the message,
 * we must insert commas, alas.
 */

fixhead(hp, tolist)
	struct header *hp;
	struct name *tolist;
{
	register int f;
	register struct name *np;

	for (f = 0, np = tolist; np != NIL; np = np->n_flink)
		if (any('@', np->n_name)) {
			f |= GCOMMA;
			break;
		}

	if (debug && f & GCOMMA)
		fprintf(stderr, "Should be inserting commas in recip lists\n");
	hp->h_to = detract(tolist, GTO|f);
	hp->h_cc = detract(tolist, GCC|f);
}
#endif NOT_USED

/*
 * Prepend a header in front of the collected stuff
 * and return the new file.
 */

FILE *
infix(hp, fi)
	struct header *hp;
	FILE *fi;
{
	extern char tempMail[];
	register FILE *nfo, *nfi;
	register int c;

	rewind(fi);
	if ((nfo = fopen(tempMail, "w")) == NULL) {
		perror(tempMail);
		return(fi);
	}
	if ((nfi = fopen(tempMail, "r")) == NULL) {
		perror(tempMail);
		fclose(nfo);
		return(fi);
	}
	remove(tempMail);
	puthead(hp, nfo, GTO|GSUBJECT|GCC|GNL);
	c = getc(fi);
	while (c != EOF) {
		putc(c, nfo);
		c = getc(fi);
	}
	if (ferror(fi)) {
		perror("read");
		return(fi);
	}
	fflush(nfo);
	if (ferror(nfo)) {
		perror(tempMail);
		fclose(nfo);
		fclose(nfi);
		return(fi);
	}
	fclose(nfo);
	fclose(fi);
	rewind(nfi);
	return(nfi);
}

/*
 * Dump the to, subject, cc header on the
 * passed file buffer.
 */

puthead(hp, fo, w)
	struct header *hp;
	FILE *fo;
{
	register int gotcha;

	gotcha = 0;
	if (hp->h_to != NOSTR && w & GTO)
		fmt("To: ", hp->h_to, fo), gotcha++;
	if (hp->h_subject != NOSTR && w & GSUBJECT)
		fprintf(fo, "Subject: %s\n", hp->h_subject), gotcha++;
	if (hp->h_cc != NOSTR && w & GCC)
		fmt("Cc: ", hp->h_cc, fo), gotcha++;
	if (hp->h_bcc != NOSTR && w & GBCC)
		fmt("Bcc: ", hp->h_bcc, fo), gotcha++;
	if (gotcha && w & GNL)
		putc('\n', fo);
}

/*
 * Format the given text to not exceed 72 characters.
 */

fmt(str, txt, fo)
	register char *str, *txt;
	register FILE *fo;
{
	register int col;
	register char *bg, *bl, *pt, ch;

	col = strlen(str);
	if (col)
		fprintf(fo, "%s", str);
	pt = bg = txt;
	bl = 0;
	while (*bg) {
		pt++;
		if (++col >72) {
			if (!bl) {
				bl = bg;
				while (*bl && !isspace(*bl))
					bl++;
			}
			if (!*bl)
				goto finish;
			ch = *bl;
			*bl = '\0';
			fprintf(fo, "%s\n    ", bg);
			col = 4;
			*bl = ch;
			pt = bg = ++bl;
			bl = 0;
		}
		if (!*pt) {
finish:
			fprintf(fo, "%s\n", bg);
			return;
		}
		if (isspace(*pt))
			bl = pt;
	}
}

/*
 * Save the outgoing mail on the passed file.
 */

savemail(name, fi)
	char name[];
	FILE *fi;
{
	register FILE *fo;
	register int c;
	time_t time();
	time_t now;
	char *ctime();
	char *n;

	if ((fo = fopen(name, "a")) == NULL) {
		perror(name);
		return;
	}
	time(&now);
	n = rflag;
	if (n == NOSTR)
		n = myname;
	fputs(delim1, fo);
	fprintf(fo, "From %s %s", n, ctime(&now));
	rewind(fi);
	for (c = getc(fi); c != EOF; c = getc(fi))
		putc(c, fo);
	fprintf(fo, "\n");
	fputs(delim2, fo);
	fflush(fo);
	if (ferror(fo))
		perror(name);
	fclose(fo);
}
